### ISSUE TITLE
Rule Change(no-unsafe-finally): Allow `return` in `finally` when `try` block dones't return anything

### ISSUE THREAD

**Issue by fisker** [eslint/eslint#18725]
### What rule do you want to change?

no-unsafe-finally

### What change do you want to make?

Generate fewer warnings

### How do you think the change should be implemented?

A new default behavior

### Example code

```js
function foo () {
  try {
    doSomething()
  } finally {
    if (bar) {
      return
    }
    doSomethingOnlyIfNotBar()
  }
}
```

### What does the rule currently do for this code?

```text
6:7 Unsafe usage of ReturnStatement.  ([no-unsafe-finally](https://eslint.org/docs/rules/no-unsafe-finally))
```

[Playground](https://eslint.org/play/#eyJ0ZXh0IjoiZnVuY3Rpb24gZm9vICgpIHtcbiAgdHJ5IHtcbiAgICBkb1NvbWV0aGluZygpXG4gIH0gZmluYWxseSB7XG4gICAgaWYgKGJhcikge1xuICAgICAgcmV0dXJuXG4gICAgfVxuICAgIGRvU29tZXRoaW5nT25seUlmTm90QmFyKClcbiAgfVxufSIsIm9wdGlvbnMiOnsicnVsZXMiOnsibm8tdW5zYWZlLWZpbmFsbHkiOlsiZXJyb3IiXX0sImxhbmd1YWdlT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6ImxhdGVzdCIsInNvdXJjZVR5cGUiOiJtb2R1bGUiLCJwYXJzZXJPcHRpb25zIjp7ImVjbWFGZWF0dXJlcyI6e319fX19)

### What will the rule do after it's changed?

No error

### Participation

- [ ] I am willing to submit a pull request to implement this change.

### Additional comments

This rule makes sense because the `return` statement changes the return value of function call, but if the function returns `void`, maybe we should allow `return` in `finally` block?

**Comment by fisker** [comment]
also includes the catch block

**Comment by Tanujkanti4441** [comment]
> also includes the catch block

Got it,

i am not sure if return should be allowed in finally block as it also overrides the exception thrown by the `try` block

```js
function foo() {
  try {
    throw new Error("error");
  } finally {
    return 3;
  }
}

try {
  const result = foo();
  console.log("result:", result)
} catch (error) {
  console.log(error);
}

// this will give 'result: 3'
```

and return in finally also ignores the return of the parent function.

```js
function foo() {
  try {
    //code
  } finally {
    return 3;
  }

return 4;
}

// foo will return 3 here
```

@eslint/eslint-team, any thoughts?

**Comment by fisker** [comment]
> it also overrides the exception thrown by the try block

I didn't know that, it makes no sense to allow `return` then.
